parquet-hadoop/src/main/java/org/apache/parquet/filter2/statisticslevel/StatisticsFilter.java

```diff
@@ -289,8 +320,14 @@ public <T extends Comparable<T>> Boolean visit(Lt<T> lt) {
 
     T value = lt.getValue();
 
-    // drop if value <= min
-    return stats.compareMinToValue(value) >= 0;
+    // we are looking for records where v < someValue
+    if (stats.compareMinToValue(value) >= 0) {
+      // drop if value <= min
+      return BLOCK_CANNOT_MATCH;
+    } else {
+      // if value > min, we must take it
+      return BLOCK_MUST_MATCH;
```

Comment:
Sometimes the min/max values are actually lower/upper bounds. Does this optimization still work for that case?

Reply:
I think  when `Statistics#hasNonNullValue` is true, it means that minMax is generated by the data content, so it can represent the real data minMax ( `Statistics#hasNonNullValue` has been processed before.)
I think if we can use minMax to judge the `BLOCK_CANNOT_MATCH` ,  we can also judge the `BLOCK_MUST_MATCH ` in some case.

Reply:
@gszadovszky Thank you for your reply!
In the original implementation, `BLOCK_MUST_MATCH` can be judged using minMax. 
So if we follow the specification, can we using minMax as the data result of an enlarged range. So we can accurately judge that when the data is not in this range.

Reply:
In fact, what I wanted to do at the beginning was to avoid the use of `BloomFilter` through `minMax` and `dictionary`(if column have) as much as possible, because the former is more accurate and `BloomFilter` may cost time and memory.

Reply:
If we allow min/max values to be lower/upper bounds and not part of the real data (as @wgtmac suggested) then there are only two scenarios we can be sure the requested value is in the row group:
* Searching for `null` and the number of nulls (is specified and) is greater than `0`
* The min value and the max value are both equal to the value we are searching for
If we expect min/max values to be part of the row groups then we can extend the related statements but it might not be a safe choice.

Dictionary filter is much more straightforward. Bloom filter should be used only if the requested value is not in the dictionary (or there is no dictionary at all) and there are pages that are not dictionary encoded. So your change clearly makes sense to differenciate the three potential results and move forward to bloom filters only in case of `BLOCK_MIGHT_MATCH`.

Reply:
@gszadovszky Thank you very much for your advice.
But this will reduce some usage scenarios. From the code, `minMax` and `dictionary` is basically the minMax of real data. But the specification doesn't say.
Can we add a config (default false, don't use `BLOCK_MUST_MATCH`) to optimize it?

Reply:
>  From the code, minMax and dictionary is basically the minMax of real data. But the specification doesn't say.

IIUC, there may be other implementation libraries where min-max is not part of the row-group data set, because the specification does not define it clearly. Although I don't know of any library that does this.

> Can we add a config (default false, don't use BLOCK_MUST_MATCH) to optimize it?

From my point of view, it is doable to introduce such a config, but I am a little concerned that it will be difficult for users to understand.

Reply:
I'm not aware of any implementation use lower/upper bounds instead of min/max values. But the specs does allow it to happen and `ColumnIndex` has stated about it [(link)](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L956) explicitly as mentioned by @gszadovszky 

I prefer to be conservative and make it easy to use:
- Regard statistics as lower/upper bounds unless `min == max`.
- Do not add any config for an aggressive optimization.

Reply:
FYI: for column index we specified it explicitly to allow shortening the min/max values (e.g. string are truncated to 64bytes by default). 
The older min/max statistics for the row groups are trickier since it is not properly specified. The issue is if only one implementation of parquet in the wild would do the similar truncate we may loose data when filtering if we use this "aggressive optimization". 
Dictionary is a completely different thing. It contains the very values from the pages themselves. It is not designed to be used for filtering (but we use anyway) but to save space.

Reply:
> Dictionary contains the very values from the pages themselves
@gszadovszky do you think it is suitable to use `BLOCK_MUST_MATCH` when all page are encoded in `DictionaryFilter`?
For example,
1. when search for data = x1 and dictionary has one element = x1  , then return `BLOCK_MUST_MATCH`
2. when search for data > x1 and every elements in dictionary > x1  , then return `BLOCK_MUST_MATCH`

Reply:
@yabola, I think you misunderstand how dictionary/filtering works. The dictionary contains all of the values which the dictionary encoded pages may contain. These pages actually not contain the values but the indices referencing to the related values in the dictionary.

So, if a searched element can be found in the dictionary you may return `BLOCK_MUST_MATCH` even if only one page is dictionary encoded. For example if the filter is `x > 1` then any element in the dictionary `> 1` would fulfill the filter so the `BLOCK_MUST_MATCH`.
If the dictionary does not contain any of the searched elements (for the previous example every elements are `<= 1`) then you may return `BLOCK_CANNOT_MATCH` only if all the related pages are dictionary encoded. Otherwise you return `BLOCK_MIGHT_MATCH` since you don't know anything about the not dictionary encoded pages based on the dictionary.

Reply:
> @yabola, I think you misunderstand how dictionary/filtering works. The dictionary contains all of the values which the dictionary encoded pages may contain. These pages actually not contain the values but the indices referencing to the related values in the dictionary.
> 
> So, if a searched element can be found in the dictionary you may return `BLOCK_MUST_MATCH` even if only one page is dictionary encoded. For example if the filter is `x > 1` then any element in the dictionary `> 1` would fulfill the filter so the `BLOCK_MUST_MATCH`. If the dictionary does not contain any of the searched elements (for the previous example every elements are `<= 1`) then you may return `BLOCK_CANNOT_MATCH` only if all the related pages are dictionary encoded. Otherwise you return `BLOCK_MIGHT_MATCH` since you don't know anything about the not dictionary encoded pages based on the dictionary.

CMIW, the `DictionaryFilter` is only enabled when all data pages are dictionary-encoded in a certain column chunk. So @yabola 's statement seems work to me.

Reply:
@wgtmac, do you refering e.g. [this](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java#L139-L143)? But this comment is only true if we are about to return `BLOCK_MIGHT_MATCH` or `BLOCK_CANNOT_MATCH`. If we want to potentially return `BLOCK_MUST_MATCH` we need to read the dictionary and check the values. So we might spend more time on dictionary filtering to potentially skip using bloom filters. But what if we don't have bloom filters? (We should not write bloom filters if all the pages are dictionary encoded anyway.)

Reply:
@gszadovszky Maybe I didn't fully understand your point. Are you concerned about the performance or correctness? The `DictionaryFilter` is already decoding the dictionary to evaluate predicates like `lt` to return `BLOCK_CANNOT_MATCH`. This patch does not introduce more overhead but supports returning `BLOCK_MUST_MATCH` at the same time.

Reply:
> We should not write bloom filters if all the pages are dictionary encoded anyway

@gszadovszky CMIW, in the existing implementation, `BloomFilter` will still generate even if all pages are encoded have dictionaries.

Reply:
I think @gszadovszky is correct.

https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileWriter.java#L884

Reply:
@wgtmac @gszadovszky I think it might be bug here because we use `PLAIN_DICTIONARY` in parquet v1? 
Bloomfilter will avoid generating only when the dictionary type is all `RLE_DICTIONARY`?
 https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileWriter.java#L884

Reply:
Good catch! I am not familiar with the old story. Does format v1 support bloom filter?

Reply:
> Good catch! I am not familiar with the old story. Does format v1 support bloom filter?

Yes, Spark use parquet v1 by default

Reply:
Godd catch indeed, @yabola! Could you open a separate jira and maybe a PR for this finding?

@wgtmac, performance. Let's see the following scenario. We have dictionary encoding but not for all the pages. We also have Bloom filter. Does it worth reading the dictionary to check if a value is in there knowing if it doesn't we still want to check the Bloom filter? I don't know the answer, maybe yes. But if it is a no, then the whole concept of this PR is questionable.
For the case of all the pages are dictionary encoded we should not have Bloom filters therefore it doesn't really matter if we return `BLOCK_MIGHT_MATCH` or `BLOCK_MUST_MATCH` in case we find the interested values in the dictionary.
Since we might already written some Bloom filters for fully dictionary encoded column chunks we should handle this scenario. But we can do it easily buy skipping reading Bloom filters in this case completely.

Reply:
> We have dictionary encoding but not for all the pages. We also have Bloom filter.

Yes, that's true.

> Does it worth reading the dictionary to check if a value is in there knowing if it doesn't we still want to check the Bloom filter?

In this case, the dictionary will not be read via `expandDictionary(meta)` by `DictionaryFilter` if `hasNonDictionaryPages(meta)` returns true and will not make performance worse. e.g. https://github.com/apache/parquet-mr/blob/261f7d2679407c833545b56f4c85a4ae8b5c9ed4/parquet-hadoop/src/main/java/org/apache/parquet/filter2/dictionarylevel/DictionaryFilter.java#L388

> For the case of all the pages are dictionary encoded we should not have Bloom filters therefore it doesn't really matter if we return BLOCK_MIGHT_MATCH or BLOCK_MUST_MATCH in case we find the interested values in the dictionary.

It is difficult to make the trade-off here. If we only have one predicate, then the dictionary will read any way, either by the DictionaryFilter or by reading the data later if the row group cannot be dropped. However, if we have other predicates that can drop the row group, then reading the dictionary here by DictionaryFilter is worthless.

Reply:
@gszadovszky @wgtmac @zhongyujiang Thank you very much for working on it.  I have some thoughts.

> We can improve (?) the case when not all the pages are dictionary encoded

1. I can't make sure if it is suitable to load dictionary even if pages are not all decoded. (I may choose not to change this behavior)

2. However considering the origin `BloomFilter` bug in parquet v1, we might have to do something to avoid using `BloomFilter`(even if pages are all encoded).  
In the code implementation we may have to use some flag to mark if dictionary `DictionaryFilter#expandDictionary` successfully (method will throw `IOException` and we can't `expandDictionary` again in `BloomFilterImpl`).
Or we could also use `BLOCK_MUST_MATCH` like this PR.

> StatisticsFilter: Because of the lower/upper bound issue we cannot really improve this (except for the specific case when min=max)

If we only use it when min=max, I think it might not worth doing it.